File: java/src/io/zihengcat/leetcode/sort/SortingTheSentence.java
package io.zihengcat.leetcode.sort;

import java.util.ArrayList;
import java.util.List;

/**
 * LeetCode 1859. Sorting the Sentence
 * https://leetcode.com/problems/sorting-the-sentence/
 */
public class SortingTheSentence {
    public static void main(String[] args) {
        // ...
    }
    public String sortSentence(String s) {
        List<String> words = split(s, ' ');
        List<String> sortedWords = new ArrayList<>(words.size());
        for (int i = 0; i < words.size(); i++) {
            sortedWords.add("");
        }
        int idx = -1;
        int wLen = 0;
        StringBuilder sBuilder = new StringBuilder();
        int n = words.size();
        for (int i = 0; i < n; i++) {
            wLen = words.get(i).length();
            idx = words.get(i).charAt(wLen - 1) - '0' - 1; // 1-indexed word position
            sortedWords.set(idx,
                sBuilder.append(words.get(i), 0, wLen - 1).toString()
            );
            sBuilder.setLength(0); // clear
        }
        return join(sortedWords, ' ');
    }
    private String join(List<String> aList, char c) {
        StringBuilder sBuilder = new StringBuilder();
        int n = aList.size();
        for (int i = 0; i < n; i++) {
            sBuilder.append(aList.get(i));
            if (i < n - 1) {
                sBuilder.append(c);
            }
        }
        return sBuilder.toString();
    }
    private List<String> split(String s, char c) {
        List<String> aList = new ArrayList<>();
        StringBuilder sBuilder = new StringBuilder();
        int startIdx = -1;
        int n = s.length();
        for (int i = 0; i < n; /* ... */) {
            startIdx = i;
            while (i < n && s.charAt(i) != c) {
                i++;
            }
            sBuilder.append(s, startIdx, i);
            aList.add(sBuilder.toString());
            sBuilder.setLength(0);
            i++;
        }
        return aList;
    }
}

/* EOF */

File: cc/sort/SortingTheSentence.cc
#include <string>
#include <vector>

using namespace std;

/**
 * LeetCode 1859. Sorting the Sentence
 * https://leetcode.com/problems/sorting-the-sentence/
 */
class SortingTheSentence {
public:
    string sortSentence(string s) {
        vector<string> words = split(s, ' ');
        vector<string> sortedWords(words.size());
        int n = words.size();
        int idx = -1;
        int wordSize = 0;
        for (int i = 0; i < n; i++) {
            wordSize = words[i].size();
            idx = words[i][wordSize - 1] - '0' - 1; // 1-indexed word position
            sortedWords[idx] = string(words[i], 0, wordSize - 1);
        }
        return join(sortedWords, " ");
    }
public:
    string join(vector<string>& v, string c) {
        string s;
        int n = v.size();
        for (int i = 0; i < n; i++) {
            s.append(v[i]);
            if (i < n - 1) {
                s.append(c);
            }
        }
        return s;
    }
    vector<string> split(string& s, char c) {
        vector<string> vec;
        int startIdx = -1;
        int n = s.size();
        for (int i = 0; i < n; /* ... */) {
            startIdx = i;
            while (i < n && s[i] != c) {
                i++;
            }
            vec.push_back(string(s, startIdx, i - startIdx));
            i++;
        }
        return vec;
    }
};

int main(int argc, char const *argv[]) {
    SortingTheSentence obj = SortingTheSentence();
    string s = "is2 sentence4 This1 a3";
    vector<string> v = obj.split(s, ' ');
    printf("%d\n", v.size());
    for (int i = 0; i < v.size(); i++) {
        printf("\"%s\"\n", v[i].c_str());
    }
    // ...
    return 0;
}

/* EOF */
